Rolling updates with drain and validate

chrislovecnm opened the pull request:
- new drain functionality with rolling updates
- using existing validate functionality with rolling-update

TODO List - Final countdown

- [x] implement feature flags
- [x] update CLI help messages with feature flag info 
- [x] update cobra docs
- [x] update docs
- [X] move TODOs to issues
- [x] drop https://github.com/kubernetes/kops/pull/1134#issuecomment-277142403 into an issue
- [X] fix drain with daemonsets
- [X] fix comments from PR review
- [X] Test with PetSets
- [X] testing more testing

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kops/1134)
<!-- Reviewable:end -->

chrislovecnm commented:
@kris-nova this is ready for a code review.  I do need MUCH testing, hence still WIP

kris-nova commented on cmd/kops/validate_cluster.go:
```diff
@@ -171,7 +171,8 @@ func RunValidateCluster(f *util.Factory, cmd *cobra.Command, args []string, out
 		// do we need to print which instance group is not ready?
 		// nodes are going to be a pain
-		fmt.Fprint(out, "\nValidation Failed\n")
+		// FIXME: format still not making sense
```
What's going on here? Do I need to get David Byrne on the horn?

chrislovecnm replied:
LOL

>David Byrne is a Scottish-born American musician who was the founding member, principal songwriter, and lead singer and guitarist of the American new wave band Talking Heads, active between 1975 and 1991.

Him ^  ... This is notes to fix the damn thing.  The output sucks.  WIP

kris-nova replied:
Mostly referring to 

> Stop making sense

I don't know why are are putting in `FIXME:`'s (is that even a thing?) in the code? What is is broken here?

chrislovecnm commented:
I am moving this out of WIP because I feel this is ready for testing and to open up to the world.  I am going to back out some work on validate, because it is pure cosmetic.  I would appreciate @justinsb or another person giving this a test drive.  I have tested primarily with private cluster, but really this is drain and validate, which is API driven and not infrastructure driven.

justinsb commented on cmd/kops/rollingupdate_cluster.go:
```diff
@@ -68,6 +70,7 @@ func init() {
 	}
 }
 
+// Run the Cobra Command - Do a rolling update for a Cluster
```
If you're going to go to the effort of adding comments, you might as well keep golint happy.  golint is a fussy little fuss pot: https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

chrislovecnm replied:
Cool

justinsb commented on upup/pkg/kutil/drain.go:
```diff
@@ -0,0 +1,656 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
I hope this fails our boilerplate check

chrislovecnm replied:
It does not ... Lol

chrislovecnm replied:
Legally it is still valid, and it is the check is from the k8s repo.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -239,25 +246,42 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
-		glog.Infof("Stopping instance %q in AWS ASG %q", instanceID, n.ASGName)
 
-		// TODO: Evacuate through k8s first?
+		drain, drainErr := NewDrainOptions(nil, u.Node.ClusterName)
```
`s/drainErr/err/g` ?

chrislovecnm replied:
There is another err in here as well that I do not want to polute. How do I do that?

kris-nova replied:
In theory you should always be able to use `err`

Although after you an implicit declaration of `err` like :

```
err := something()
``` 

You no longer need the `:` to re assign the `err` to a new value. Go is **not** immutable like our scala neighbors, so this is perfectly fine. In other words, once `err` has been assigned via `err := something()` you now only need to do :

```
err = somethingElse()
```

in order to keep using `err`! 

There are a few caveat's.. For instance if you had  `err` used above in your code, **BUT** you were **ALSO** declaring a **NEW** variable that was returned in a tuple from something as in:

```
n, err := something()
```

You could get away with using a `:` even though `err` has already been defined. 

Go ¯ \\ _ ( ツ ) _ / ¯

chrislovecnm replied:
OMG ... I have a novel to read.... Let me get these changes in, and I will read this.  Thanks!!!!

chrislovecnm replied:
So in theory.  If you absolutely want me to do this I can.  Frankly, it makes my head hurt.  Do I need to do this?

We have some tricky logic situations, where I could have an `err` flow from one node to another.  I would have to carefully scope the `err`.  So carefully scope or have 3 variables instead of one variable name, and still actually still have three variables in mem?

What is the simplest this to do?

kris-nova replied:
You will get a LGTM if you leave it as is - albeit remind yourself the `G` means `good` and not `great`

chrislovecnm replied:
:facepalm

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -239,25 +246,42 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+					glog.Warningf("Drain failed %s", u.Node.Name)
+				}
+		  }
+		}
```
Perhaps pull this block out into a `drainNode`, and then just log something like `glog.Warningf("error draining node %s (instance %s); will terminate anyway")`

chrislovecnm replied:
Great idea. Will refactor.

chrislovecnm replied:
Refactored

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -266,8 +290,26 @@ func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval
 		time.Sleep(interval)
+
+		// Wait until the cluster is happy
+		// TODO: do we need to respect cloud only??
```
Probably yes.... it's really important in the case when you have pushed a bad config and can't reach the API

chrislovecnm replied:
That was your note. Please explain

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -266,8 +290,26 @@ func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval
+			_, validateDidNotPass := validate.ValidateCluster(u.Node.ClusterName, instanceGroupList, k8sClient)
+
+			if validateDidNotPass != nil {
+				glog.V(2).Infof("Unable to validate kuberetes cluster %s, %v.", u.Node.ClusterName, validateDidNotPass)
```
`s/kuberetes/kubernetes/g`

chrislovecnm replied:
Engrish hard

chrislovecnm replied:
Fixed

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -266,8 +290,26 @@ func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval
+		// TODO: do we need to respect cloud only??
+		for i := 0; i < retries; i++ {
+
+			_, validateDidNotPass := validate.ValidateCluster(u.Node.ClusterName, instanceGroupList, k8sClient)
```
Are we validating the cluster or the node here?

chrislovecnm replied:
The cluster... The code does not know which node is new... Yay asg. A node can be cordoned, and it is a valid cluster. The new node has to be schedulable.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -266,8 +290,26 @@ func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval
+			}
+		}
+
+		// TODO: what do we do if validation fails?
```
Presumably stop, unless we're running in `--force` (or another flag if we think that flag is orthogonal)

chrislovecnm replied:
Agreed

chrislovecnm replied:
I added flags for failing on drain and failing on validate.

justinsb commented on upup/pkg/kutil/drain.go:
```diff
@@ -0,0 +1,656 @@
+////
+// Based off of drain in kubectl
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/drain.go
+////
```
Well this is incredibly painful.

Can we import the task instead, and use the code that way? For example calling `DrainOptions.RunCordonOrUncordon` directly?  What changes did you have to make?

chrislovecnm replied:
There code is lacking ... Removed dependencies, cleaned up crap. Only have some way to do it, fixed unit testing, and probably have 3 edge cases. Need to document.

kris-nova replied:
Their*

kris-nova commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -239,25 +246,42 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+		} else {
+			drainErr = drain.SetupDrain(u.Node.Name)
+			if drainErr != nil {
+				glog.Warningf("setupErr: %v", drainErr)
```
Do we `glog.Warningf` anywhere else? Also why are we putting `setupErr` in a warning message? What does that even mean?

chrislovecnm replied:
Maybe I should use actual English ;)

kris-nova commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -266,8 +290,26 @@ func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval
+
+		// Wait until the cluster is happy
+		// TODO: do we need to respect cloud only??
+		for i := 0; i < retries; i++ {
```
```
for i := 0; i <= retries; i++ {
```

LTE <=

chrislovecnm replied:
Fixed

kris-nova commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -239,25 +246,42 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+
+				if drainErr != nil {
+					// TODO: fail here??
+					glog.Warningf("Drain failed %s", u.Node.Name)
```
I don't the consequence of a drain failing - it sounds like the entire operation should fail? I don't know - concurrent failures are terrifying to be honest - I would need to know more about what we are attempting to do here, and what `drain` means in k8s (to you)

chrislovecnm replied:
Yah we need to talk. I need to test more for edge cases, but my thought is we are ha, and we need to upgrade.  We can fail and see how it goes, but a lot of people use --force.

chrislovecnm replied:
Maybe we should support both ... Hrmmmm.

chrislovecnm replied:
So I implemented flags to fail on drain error and fail on validate.  I need another to force a drain or be nice :)

chrislovecnm commented:
Still need to test, and add a flag to not force a drain.  Not sure how I am going to check it ...

chrislovecnm commented:
@kris-nova / @justinsb ~ address all of you PR concerns.  Can I get a final call on `err` debate 🙊 

Need to test more and thing about another flag for config.  Probably should get this on the agenda tomorrow.

chrislovecnm commented:
TODO:
- test
- update docs
- tweak var names
- file issue on kubectl drain

As I thought we should not allow the user to not use `--force`.  In one of our user's words.  It is useless without force ... Hence use the force ...

hubt commented:
This will be nice and useful.

While you're working in this code, it looks like rolling update only parallelizes by rolling one node from each instance group concurrently. In a medium or large cluster, that's going to take a while. Rather than having to wait for an hour or temporarily expand the ASG, it'd be nice if I could specify a roll concurrency parameter in absolute number or as a percentage, similar to how a `rollingUpdateStrategy` of a deployment has a `maxUnavailable`. It's true that in a heterogeneous cluster it could simultaneously roll all your big nodes, but that's an acceptable risk to save hours of waiting.

If you want to be fancy, you could attempt to auto-detect an acceptable roll rate by making sure there are no `Pending` pods, but that's probably too tricky for right now.

chrislovecnm commented:
Need to test this https://github.com/kubernetes/kops/issues/1009

jrnt30 commented:
I see that this was taken out of WIP however contains several TODO and FIXME which made sense to me (specifically the flag name change).  

Are these things still going to be addressed?  If not, should the comments then be removed since it would not be as feasible to change after it's merged in?

chrislovecnm commented:
@jrnt30 this is still WIP ...

chrislovecnm commented:
@jrnt30 I have even more work to do with the api changes dropped in.  Now bastions are instancegroups we cannot drain them :)  So still a work in progress.

jrnt30 commented:
No worries, thanks for the work and updates. Interested in how it progresses

chrislovecnm commented:
@jrnt30 thanks ... if once I get the changes in if you mind testing, and providing feedback.  That would be AWESOME.  This is a critical update for the community.

chrislovecnm commented:
@hubt I agree, just lets get through that in another PR.  Opening an issue with your ideas :)

AlexLast commented:
It would be great if this could be done, I just tried a rolling update on a cluster with 3 nodes and there was a 1-2 minute period where there were no ready nodes e.g no running pods.

Might be worth waiting for 2/2 EC2 status checks to report good before terminating the next node in sequence.

chrislovecnm commented:
@AlexLast still WIP ... going through the k8s upgrade now ;(  Drain test is broken ... yay

chrislovecnm commented:
Moved this out of WIP, but also into the backlog.

chrislovecnm commented:
To the people tracking this issue.  Testing would be appreciated, and we are on the verge of a 1.5.0 release, and will merge into head, after the 1.5.0 release.

ese commented:
I tried to test this merged with b419f2013df4986f80c36f663a04fc58c1dd7f84 (master) but `kops rolling-update cluster` ends inmediatly after run with no output

chrislovecnm commented:
@ese you add an ig?  Make a cluster change?

1. kops edit <- bump the k8s version or downgrade
2. kops update --yes
3. kops rolling-update

If we are not saying anything and just exit, we need to file that as a bug.  I appreciate your patience if I missed something :)

ese commented:
@chrislovecnm Yes I changed serveral values and other versions of kops detect changes to apply with rolling-update

chrislovecnm commented:
@ese thanks - but it back into WIP. Gotta make this awesome ;)

nicolasbelanger commented:
Worth mentioning that we just had a flawless rolling-upgrade here!  Thank you guys!

chrislovecnm commented:
@ese @nicolasbelanger @AlexLast @jrnt30 @justinsb @kris-nova 

If anyone has time, testing is appreciated.  I want to do some testing with a real cluster with apps running ;)  But this is looking stable

chrislovecnm commented:
Oh and 1.4.8 -> 1.5.2 with petset ... yep not good.  The rolling-update deletes the pods ;)

@justinsb I am wondering if I should check for non-backwards compatible components such as PetSet to Stateful set .... Coming to get complicated really fast :(  My changes have not created this problem.  We have always had it

chrislovecnm commented:
@hubt did we get yo ideas about || updates into an issue?  I think we need to increase the ig size before we look at || rolling updates.  If we are doing two nodes at the same time, we need two new nodes before.  Basically, I need to ensure that we don't run out of headroom in the cluster.

justinsb commented on hack/dev-build.sh:
```diff
@@ -122,6 +122,12 @@ kops create cluster \
+  --networking $NETWORKING \
+  --bastion="true" \
+=======
+>>>>>>> master
```
Rebase problem

chrislovecnm replied:
Geeze damn file snuck in, should not even be in the merge

chrislovecnm commented:
@justinsb backed out changes to the dev-build.sh file.

ese commented:
Created a cluster with alpha channel and changed k8s version to 1.5.1
`kops.test create cluster --name demerzel.aurora.ag  --node-count 2 --topology private --node-size m4.large --bastion --ssh-public-key ~/.ssh/kubernetes/demerzel.aurora.ag.pub --channel alpha --zones eu-west-1a,eu-west-1b,eu-west-1c --networking calico`

Changing k8s version to 1.5.2 and rolling-update 

After upgrade master got this.

```
-> kops.test rolling-update cluster --yes 
Using cluster from kubectl context: demerzel.aurora.ag

NAME			STATUS		NEEDUPDATE	READY	MIN	MAX	NODES
bastions		Ready		0		1	1	1	0
master-eu-west-1a	Ready		0		1	1	1	1
nodes			NeedsUpdate	2		0	2	2	2
panic: runtime error: index out of range

goroutine 189 [running]:
panic(0x1bc5060, 0xc42000e150)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
k8s.io/kops/upup/pkg/kutil.(*CloudInstanceGroup).RollingUpdate(0xc420c02cc0, 0xc420027680, 0xc4200276c0, 0xc420027660)
	/go/src/k8s.io/kops/upup/pkg/kutil/rollingupdate_cluster.go:347 +0xc2e
k8s.io/kops/upup/pkg/kutil.(*RollingUpdateCluster).RollingUpdate.func2(0xc420bff890, 0xc4203c46a0, 0xc420bff860, 0xc4203c4700, 0xc420b9ef00, 0xc420ad10e0)
	/go/src/k8s.io/kops/upup/pkg/kutil/rollingupdate_cluster.go:218 +0x2f7
created by k8s.io/kops/upup/pkg/kutil.(*RollingUpdateCluster).RollingUpdate
	/go/src/k8s.io/kops/upup/pkg/kutil/rollingupdate_cluster.go:225 +0x56d
```

hubt commented:
I created issue #1718 for the changes I suggested

ese commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,24 +340,56 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+	if !rollingUpdateData.IsBastion && rollingUpdateData.FailOnValidate && !rollingUpdateData.CloudOnly {
+		_, err := validate.ValidateCluster(rollingUpdateData.ClusterName, rollingUpdateData.InstanceGroupList, rollingUpdateData.K8sClient)
+		if err != nil {
+			return fmt.Errorf("Cluster %s does not pass validateion", rollingUpdateData.ClusterName)
```
typo *validateion

kris-nova replied:
This is a deal breaker.. we need this fixed

ese commented:
I would use `delete-local-data` option for drain when `force` is true

chrislovecnm commented:
@ese since we are rolling the instance completely, do we need to delete the local data?  New instances completely.

ese commented:
@chrislovecnm this is for get pods with local data evicted before stop the instances

chrislovecnm commented:
@ese thanks a ton.  So interesting thing there is a bug with daemonsets then.  Let me create a note and will get this updated.

kris-nova commented:
@chrislovecnm can we get a squash please?

chrislovecnm commented:
@kris-nova the Togo is a deal breaker? Please provide more context.

chrislovecnm commented:
I need to refactor this into a feature flag, so the code is ready to test, but it is wip

chrislovecnm commented:
@kris-nova we are going to need to squash when we merge into master.  I am getting

```
$ grb -i master
error: could not apply c6263bb... start on rolling update

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply c6263bbda9eeeea1075824dfc2f15ea9f1de46c4... start on rolling update
```

The rebase is really nasty, and I would rather not go through it a third time.

kris-nova commented:
Hey @chrislovecnm,

So using `git` is a fundamental part of contributing to the project. We expect all contributors to be able to manage a simple merge conflict/rebase. Here is a great [article on how to handle merge conflicts from the command line](https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/). 

19 commits for a single PR is a bit much, can we get this squashed so the logs are more manageable?

kris-nova commented on cmd/kops/rollingupdatecluster.go:
```diff
@@ -35,11 +37,14 @@ import (
 	CloudOnly bool
 
+	ForceDrain      bool
+	FailOnValidate  bool
```
Wondering if this is a false positive? Can we make the name a little more verbose here?

```
FailUpdateOnFailedValidation
```

Just so the user is sure about the logic?

chrislovecnm replied:
I am going to make sure the cli instruction is up to date as well.

chrislovecnm replied:
Working on name change will be pushed soon

kris-nova commented on cmd/kops/rollingupdatecluster.go:
```diff
@@ -137,6 +150,7 @@ func RunRollingUpdateCluster(f *util.Factory, out io.Writer, options *RollingUpd
 			return fmt.Errorf("cannot build kube client for %q: %v", contextName, err)
 		}
 
+		// TODO use k8s client here so that we can be backwards compat
```
So we need to start respecting https://github.com/kubernetes/kops/issues/1262

Our comments are moving towards documentation. So we need to focus on things like complete (and correctly spelled) sentences. With correct punctuation, etc.. 

Please update this line so that it reads correctly.

chrislovecnm replied:
I am going to move the TODO's to issues.

kris-nova commented on cmd/kops/rollingupdatecluster.go:
```diff
@@ -264,5 +271,27 @@ func RunRollingUpdateCluster(f *util.Factory, out io.Writer, options *RollingUpd
+			ClusterName:    options.ClusterName,
+			Cloud:          cloud,
+		}
+		glog.V(2).Infof("New rolling update with drain and validate enabled")
```
I think the convention in kops is to keep log lines and errors lower cased.

sstarcher replied:
Golang convention is starts with lowercase and no punctuation.

chrislovecnm replied:
I will look at what the rest of the code is doing, and match that.  We may be doing it incorrectly, but it would be better to do it consistently.

kris-nova commented on pkg/featureflag/featureflag.go:
```diff
@@ -38,6 +38,9 @@ func Bool(b bool) *bool {
 var DNSPreCreate = New("DNSPreCreate", Bool(true))
 
+// if set will use new rolling update code that will drain and validate
+var ValidiateAndDrainRollingUpdate = New("ValidiateAndDrainRollingUpdate", Bool(true))
```
Why are we `New()`ing a bool here? Just curious as to the use case.

kris-nova commented on upup/pkg/kutil/drain.go:
```diff
@@ -0,0 +1,674 @@
+/*
```
Why are we adding more code to kutil? This package seems a bit misc'y enough as it is.. wondering if there is a better home for this.

kris-nova commented on upup/pkg/kutil/drain.go:
```diff
@@ -0,0 +1,674 @@
+
+package kutil
+
+////
```
What's going on here? Why do we have `////`?

chrislovecnm replied:
It is the way that I mark comments, habit from working on Terraform code.  Also I don't think we really have a style guide, does core have one?

I will pull.

kris-nova commented on upup/pkg/kutil/drain.go:
```diff
@@ -0,0 +1,674 @@
+// TODO: refactor our client to be like this client
+
+// FIXME: look at 1.5 refactor
+// FIXME: we are deleting local storage for daemon sets, and why even delete local storage??
```
Proper punctuation please

kris-nova commented on upup/pkg/kutil/drain.go:
```diff
@@ -0,0 +1,674 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Why are we copy/pasting code from the main repo into kops? Can we just reference the library directly so we only have one source of truth? This seems a bit worrisome that we are attempting to import an almost verbatim copy of [drain.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/drain.go) into kops.. Whats the advantage? And why `kutil`?

chrislovecnm replied:
@kris-nova I have already found two bugs in kubectl.  The code from kubectl may not be robust enough.  Moreover, I am planning on using the API client, in other iterations, as we need to support 1.4.x and 1.5.x.  For instance, 1.4.x kubectl drain fails if you are running a Petset.  Also, this code is removing the local storage of a daemon set, which is not quite correct.  DSs should not be touched.

I would like to work on improving this after this iteration, as this is functioning well currently.

justinsb replied:
Can you paste the issues for the kubectl bugs?  I struggle to understand why we would do this as anything other than a short-lived fork, until our fixes made it upstream.

kris-nova commented:
Can you clarify what you mean here? I just want to make sure I understand you correctly..

> we are going to need to squash when we merge into master.

Are you suggesting we squash commits and then do a force push onto the master branch?

sstarcher commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,319 @@
+)
+
+// RollingUpdateCluster restarts cluster nodes
+type RollingUpdateClusterDV struct {
```
First value in comment should be the same as the struct name.  What's DV?

chrislovecnm replied:
That was part of the refactor, to use the feature flag.  I can put the files in separate folders, or actually spell.  I think I will refactor and spell `DrainAndValidate` :)

sstarcher commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,319 @@
+	ClusterName string
+}
+
+// RollingUpdateData is used to pass information to perform a rolling update
```
same as above

sstarcher commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,319 @@
+const retries = 8
+
+// TODO: should we check to see if api updates exist in the cluster
+// TODO: for instance should we check if Petsets exist when upgrading 1.4.x -> 1.5.x
```
To be safe on the user you could add a simple abort if petsets exist and you are upgrading to 1.5, but handling more than just a simple abort would be complicated.

chrislovecnm replied:
I will vet this design with @bprashanth

chrislovecnm replied:
@sstarcher we need to switch to the go client as well, to get this functionality.  I am using the kubernetes api, and we need to switch to the go client.

chrislovecnm commented:
So here is the word on Petsets.  We need another iteration on this to improve for petsets.  We can roll with this now, but it has the opportunity to be disruptive for Pets or States ... StatefulSets

>  i think the idea was to achieve it through PDB (pod disruption budget), so your normal deletes/drains would work the same way always, and respect PDB, which would probably default to "never delete more than 1 pod at a time" for petset. Suggest asking for status on https://github.com/kubernetes/kubernetes/issues/33727, since it's on the kubectl team to finish that up.

Technically you should only delete one pet at a time and let the petset normalize.  A well designed petset deployment utilized anit-affinity rule to ensure that multiple pets are on the same node.  Again this is complicated and needs to mature over time.  I appreciate @justinsb having me put it in a feature flag.  I am making a note to put this into an issue

bprashanth commented:
I think @foxish and/or @ymqytw might have context on the status of drain and petset

chrislovecnm commented:
@kris-nova should have addressed your PR review comments.

Everyone that is following along.  Doing more testing, and would like to address the issue I found with deleting daemon sets with local storage.

justinsb commented on docs/cli/kops_rolling-update_cluster.md:
````diff
@@ -20,6 +23,8 @@ kops rolling-update cluster
 ```
       --bastion-interval duration    Time to wait between restarting bastions (default 5m0s)
       --cloudonly                    Perform rolling update without confirming progress with k8s
+      --fail-on-drain-error          The rolling-update will fail if draining a node fails. Enable with KOPS_FEATURE_FLAGS='+ValidiateAndDrainRollingUpdate'
````
Do we expect users to tweak these flags?  Are these effectively the equivalent of the `ValidiateAndDrainRollingUpdate` feature flag?

chrislovecnm replied:
Not really.  It allows for the rolling update to 'try' to validate, but not bail.  For instance, if you have one node that is not happy or a master that is not happy, you can still do a rolling update, and attempt to validate.  `--cloudonly` turns off everything.

Do I need to wordsmith the help docs?

justinsb replied:
If we don't expect users to change the flags, I'm thinking we probably don't want/need them.  It isn't clear why users would change the flags (vs disabling drain)

chrislovecnm replied:
I need to make it more clear.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,313 @@
+// CreateRollingUpdateData creates a RollingUpdateClusterDrainValidate struct.
+func (c *RollingUpdateClusterDrainValidate) CreateRollingUpdateData(instanceGroups *api.InstanceGroupList, isBastion bool) *RollingUpdateDataDrainValidate {
+	return &RollingUpdateDataDrainValidate{
+		Cloud:             c.Cloud,
```
This is a little unusual... why not include a field with the *RollingUpdateClusterDrainValidate, and avoid duplicating the fields?

chrislovecnm replied:
we can ... just wanted a clean data structure transition, so that we don't pass in the bastion flag incorrectly.  I can refactor to just pass in the flag as well.  You want me to refactor now or later?

chrislovecnm replied:
Also, it did not start out that way ... kinda ended up in that direction.  Same question about the two refactors you asked for.  Can we push them into another PR?  Completely fair for you to say no ...

justinsb replied:
I am fine with leaving the field duplication here - it's just a style question.  The other concerns create more technical debt so I think we should address them now.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,313 @@
+	}
+
+	if rollingUpdateData.InstanceGroupList == nil {
+		return fmt.Errorf("rollingUpdate is missing a the InstanceGroupList")
```
typo

chrislovecnm replied:
pushing fix soon

justinsb commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,313 @@
+
+	for _, u := range update {
+
+		if !rollingUpdateData.IsBastion {
```
I think we could probably split this into a function `drainTheNode`

chrislovecnm replied:
So this is when I split up code.  Let me see how this sounds. 

If you need to test something specifically, I break it out.  
If you need to call the same lines of code from different functions, I break it out.

What value do we have breaking it out here?

justinsb replied:
Most trivially it'll avoid the nested if blocks.  It also keeps the function more readable IMO.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,313 @@
+
+			// Wait until the cluster is happy
+			for i := 0; i <= retries; i++ {
+				_, err = validate.ValidateCluster(rollingUpdateData.ClusterName, rollingUpdateData.InstanceGroupList, rollingUpdateData.K8sClient)
```
Does this wait for the new node?

chrislovecnm replied:
Absolutely, it waits for the new node to be in service.  Once @sstarcher changes get in, we will wait for kube-system to stabilize as well.

justinsb replied:
I think it checks that all nodes are ready.  But it doesn't check that the new node has registered with the API.  If the rolling-update meant that all new instances failed to register with the API server, would we stop the rolling-update?

justinsb commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,313 @@
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+	api "k8s.io/kops/pkg/apis/kops"
+	validate "k8s.io/kops/pkg/validation"
```
Any reason to alias validation?

chrislovecnm replied:
nopes ... changing soon

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -276,16 +276,11 @@ func (n *CloudInstanceGroup) RollingUpdate(cloud fi.Cloud, force bool, interval
 
-		// TODO: Evacuate through k8s first?
-
-		// TODO: Temporarily increase size of ASG?
```
let's leave these comments - I think the only addresses the "Evacuate through k8s first" comment

chrislovecnm replied:
👍 - I was thinking of backing out all changes to this file.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster_validate_drain.go:
```diff
@@ -0,0 +1,313 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
The logic here is not that different to normal rollingupdate.  I feel like it boils down to whether we run validate, and whether we call drain, which could just be flags?  I don't think we need to copy-pasta this.

chrislovecnm replied:
The problem is the structs.  I have modified the data structures.  Yes, I would do that, but there is less risk to the code with a copy-pasta.  ... Y/N?

justinsb replied:
I think this is:

* Add two fields `ShouldDrain` and `ShouldValidate`
* Call two (new) `drain` and `validate`

That's what it looked like the diff would be - if that's the case that is very acceptable risk IMO.

foxish commented:
> I think @foxish and/or @ymqytw might have context on the status of drain 

For 1.5, kubectl drain was changed to use the eviction endpoint which respects PDBs.
There was a long discussion on defaulting in https://github.com/kubernetes/kubernetes/issues/35318, but no conclusions yet.

chrislovecnm commented:
@justinsb you were asking about deleting local data.  https://github.com/kubernetes/kops/pull/1134#issuecomment-276461923 @ese explained it

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -98,9 +108,11 @@ func FindCloudInstanceGroups(cloud fi.Cloud, cluster *api.Cluster, instancegroup
+
 	if len(groups) == 0 {
-		return nil
+		return fmt.Errorf("groups length cannot be zero")
```
Any reason to make this an error?

chrislovecnm replied:
I will change back

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -118,7 +130,7 @@ func (c *RollingUpdateCluster) RollingUpdate(groups map[string]*CloudInstanceGro
 			bastionGroups[k] = group
 		default:
-			return fmt.Errorf("unknown group type for group %q", group.InstanceGroup.ObjectMeta.Name)
+			return fmt.Errorf("unknown group type for group: %q", group.InstanceGroup.ObjectMeta.Name)
```
I don't think we need the colon here.  If you wanted to print the type - which is not a bad idea - then we would.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
 	}
+
+	if isBastion || rollingUpdateData.CloudOnly {
+		glog.Infof("Not validating the cluster")
```
I think that is going to be a confusing message when it is printed when you asked for validation.

chrislovecnm replied:
Will need two different log messages or just drop logging.

justinsb commented on cmd/kops/rollingupdatecluster.go:
```diff
@@ -82,9 +105,12 @@ To perform rolling update, you need to update the cloud resources first with "ko
 
+	cmd.Flags().BoolVar(&options.FailOnDrainError, "fail-on-drain-error", false, "The rolling-update will fail if draining a node fails. Enable with KOPS_FEATURE_FLAGS='+DrainAndValidateRollingUpdate'")
+	cmd.Flags().BoolVar(&options.FailOnValidate, "fail-on-validate", true, "The rolling-update will fail if the cluster fails to validate. Enable with KOPS_FEATURE_FLAGS='+DrainAndValidateRollingUpdate'")
+	cmd.Flags().IntVar(&options.ValidateRetries, "validate-retries", 8, "The rolling-update will fail if the cluster fails to validate within the number of retries. Enable with KOPS_FEATURE_FLAGS='+DrainAndValidateRollingUpdate'")
```
Probably this should be a duration?  The number of retries doesn't mean anything to a user.

I would probably not expose it initially.

kris-nova replied:
+1 duration

chrislovecnm replied:
We would need number of retries and duration.  I am looping for a count and sleeping for a period.

chrislovecnm replied:
Also, this will be the third time that I have changed these flags.  I would like this to be the last change.

chrislovecnm replied:
I updated the message to explain more.

justinsb commented on cmd/kops/rollingupdatecluster.go:
```diff
@@ -82,9 +105,12 @@ To perform rolling update, you need to update the cloud resources first with "ko
 	cmd.Flags().StringSliceVar(&options.InstanceGroups, "instance-group", options.InstanceGroups, "List of instance groups to update (defaults to all if not specified)")
 
+	cmd.Flags().BoolVar(&options.FailOnDrainError, "fail-on-drain-error", false, "The rolling-update will fail if draining a node fails. Enable with KOPS_FEATURE_FLAGS='+DrainAndValidateRollingUpdate'")
+	cmd.Flags().BoolVar(&options.FailOnValidate, "fail-on-validate", true, "The rolling-update will fail if the cluster fails to validate. Enable with KOPS_FEATURE_FLAGS='+DrainAndValidateRollingUpdate'")
```
It's weird to have two flags "fail-on-drain-error" and "fail-on-validate".  Presumably "fail-on-validate-error"?

chrislovecnm replied:
Validate is one step, drain is another.  There is a use case where I don't care if a drain succeeds, but I need validate to work.

justinsb replied:
I was unclear: Can we rename `fail-on-validate` -> `fail-on-validate-error` ?

chrislovecnm replied:
That we can do!

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			glog.Warningf("Not draining nodes - cloud only is set.")
+		} else if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+
+			glog.Infof("Draining the node.")
```
Probably should print the node name.

chrislovecnm replied:
Will update

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			if err = n.DeleteAWSInstance(u, c); err != nil {
+				return err
+			}
+			glog.Infof("Updated a bastion instance and continuing.")
```
Probably should include the instance name.  Also, the message is misleading - it wasn't "updated"....

chrislovecnm replied:
Will update

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			glog.Infof("Draining the node.")
+
+			if err = n.DrainNode(u, rollingUpdateData); err == nil {
+				glog.Errorf("Error draining node: %s.", err)
```
You should just `return fmt.Errorf("error draining node %q: %v", nodeName, err)`

chrislovecnm replied:
Will update

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+
+		if err = n.DeleteAWSInstance(u, c); err != nil {
+			glog.Errorf("Error deleting aws instance: %s.", err)
+			return err
```
`return fmt.Errorf("error deleting AWS instance %q: %v"...`

chrislovecnm replied:
Will update

chrislovecnm replied:
changed a bunch

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+		if isBastion {
+
+			if err = n.DeleteAWSInstance(u, c); err != nil {
+				return err
```
Probably `return fmt.Errorf("error deleting AWS instance %q: %v"...` for consistency

chrislovecnm replied:
Will update

chrislovecnm replied:
We are getting that exact error from the method that is called.  I would rather not duplicate.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			return nil
+		} else if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+
+			glog.Infof("Validating the cluster.")
```
So this is all super-chatty.  I wonder if we should take most of these to V(2)?

chrislovecnm replied:
I like the chatty, because I am watching the cluster update.  Makes me feel confident that kops is doing its job.

Thoughts?

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			glog.Infof("Validating the cluster.")
+			if err = n.ValidateClusterWithRetries(rollingUpdateData, instanceGroupList, t); err != nil {
+				glog.Errorf("Error validating cluster: %s.", err)
+				return err
```
`return fmt.Errorf("error validating cluster after removing instance %q: %v"...`

chrislovecnm replied:
👍

chrislovecnm replied:
Actually removed it as it is not an error if FailOnValidate is false

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			}
+
+			if err != nil && rollingUpdateData.FailOnValidate {
+				return fmt.Errorf("validation timed out while performing rolling update: %v", err)
```
Unreachable?

chrislovecnm replied:
Logic error ... I think I fixed it ... please double check.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+
+	_, err := validation.ValidateCluster(rollingUpdateData.ClusterName, instanceGroupList, rollingUpdateData.K8sClient)
+	if err != nil {
+		return fmt.Errorf("cluster %s does not pass validation", rollingUpdateData.ClusterName)
```
Include the inner error?  Also I don't think anyone is going to need the cluster name?

chrislovecnm replied:
I can include the inner error, we may need to reword the inner error.

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
 
-		request := &ec2.TerminateInstancesInput{
-			InstanceIds: []*string{u.ASGInstance.InstanceId},
+		glog.Warningf("Error creating drain: %v.", err)
```
`Warningf("Error draining node %q: %v"...`

chrislovecnm replied:
👍

chrislovecnm replied:
It is actually a cluster api error on a node error.

chrislovecnm replied:
Updated error message

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+			return fmt.Errorf("error creating drain: %v.", err)
+		} else {
+			glog.Infof("Proceeding with rolling-update since fail-on-drain-error is set to false.")
 		}
```
You can just `return nil` here, to save the nesting

chrislovecnm replied:
More context please

chrislovecnm replied:
The else statement is just doing logging.  I can chain the if statements together, but I personally I mess up and don't like reading long if statements.

chrislovecnm replied:
Figured it out :)

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+		err = drain.DrainTheNode(u.Node.Name)
 		if err != nil {
-			return fmt.Errorf("error deleting instance %q: %v", instanceID, err)
+			glog.Warningf("Error draining node: %v.", err)
```
Should we put the node name in this message?

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
-		glog.Infof("Stopping instance %q in AWS ASG %q", instanceID, n.ASGName)
 
-		// TODO: Evacuate through k8s first?
+		if isBastion {
```
Can we just add this block to the conditions in the next if statement, where we don't drain the instance?

justinsb commented on upup/pkg/kutil/rollingupdate_cluster.go:
```diff
@@ -269,35 +281,169 @@ func buildCloudInstanceGroup(ig *api.InstanceGroup, g *autoscaling.Group, nodeMa
+
+		if rollingUpdateData.CloudOnly {
+			glog.Warningf("Not validating nodes as cloudonly flag is set.")
+			return nil
```
Are we exiting the for-loop here?

chrislovecnm replied:
Thanks, changed to continue

chrislovecnm commented:
@justinsb and @kris-nova update done, and I need to test again.  I think the flags are the only concern that we have open.   

The logging is a bit more verbose than I think @justinsb wants, but as a user, it gives me visual re-assuring that the update is proceeding.

chrislovecnm commented:
Added WIP, as I am seeing an odd behavior in testing.

chrislovecnm commented:
@k8s-bot aws e2e test this

chrislovecnm commented:
Moved this back to WIP, because it depends on https://github.com/kubernetes/kops/pull/2014

justinsb commented on cmd/kops/rollingupdatecluster.go:
```diff
@@ -82,9 +111,15 @@ To perform rolling update, you need to update the cloud resources first with "ko
 	cmd.Flags().StringSliceVar(&options.InstanceGroups, "instance-group", options.InstanceGroups, "List of instance groups to update (defaults to all if not specified)")
 
+	if featureflag.DrainAndValidateRollingUpdate.Enabled() {
+		cmd.Flags().BoolVar(&options.FailOnDrainError, "fail-on-drain-error", true, "The rolling-update will fail if draining a node fails.")
```
Don't use `true`, use options.FailOnDrainError

brandoncole commented:
LGTM

sstarcher commented:
LGTM
